posthog/hogql/test/test_printer.py

```diff
@@ -352,6 +361,241 @@ def test_property_groups(self):
+        # not. We can still utilize the bloom filter index on keys, but the empty string isn't stored in the bloom
+        # filter so it won't be used here.
+        self._test_property_group_comparison(
+            "properties.key = ''",
+            "and(has(events.properties_group_custom, %(hogql_val_0)s), equals(events.properties_group_custom[%(hogql_val_0)s], %(hogql_val_1)s))",
+            {"hogql_val_0": "key", "hogql_val_1": ""},
+            expected_skip_indexes_used={"properties_group_custom_keys_bf"},
+        )
+        self._test_property_group_comparison(
+            "equals(properties.key, '')",
+            "and(has(events.properties_group_custom, %(hogql_val_0)s), equals(events.properties_group_custom[%(hogql_val_0)s], %(hogql_val_1)s))",
+            {"hogql_val_0": "key", "hogql_val_1": ""},
+            expected_skip_indexes_used={"properties_group_custom_keys_bf"},
+        )
+
+    def test_property_groups_optimized_null_comparisons(self) -> None:
+        # NOT NULL comparisons should check to see if the key exists within the map (and should use the bloom filter to
+        # optimize the check), but do not need to load the values subcolumn.
+        self._test_property_group_comparison(
+            "properties.key is not null",
+            "has(events.properties_group_custom, %(hogql_val_0)s)",
+            {"hogql_val_0": "key"},
+            expected_skip_indexes_used={"properties_group_custom_keys_bf"},
+        )
+        self._test_property_group_comparison(
+            "properties.key != null",
+            "has(events.properties_group_custom, %(hogql_val_0)s)",
+            {"hogql_val_0": "key"},
+            expected_skip_indexes_used={"properties_group_custom_keys_bf"},
+        )
+        self._test_property_group_comparison(
+            "isNotNull(properties.key)",
+            "has(events.properties_group_custom, %(hogql_val_0)s)",
+            {"hogql_val_0": "key"},
+            expected_skip_indexes_used={"properties_group_custom_keys_bf"},
+        )
+
+        # NULL comparisons don't really benefit from the bloom filter index like NOT NULL comparisons do, but like
+        # above, only need to check the keys subcolumn and not the values subcolumn.
+        self._test_property_group_comparison(
+            "properties.key is null",
+            "not(has(events.properties_group_custom, %(hogql_val_0)s))",
+            {"hogql_val_0": "key"},
+        )
+        self._test_property_group_comparison(
+            "properties.key = null",
+            "not(has(events.properties_group_custom, %(hogql_val_0)s))",
+            {"hogql_val_0": "key"},
+        )
+        self._test_property_group_comparison(
+            "isNull(properties.key)",
+            "not(has(events.properties_group_custom, %(hogql_val_0)s))",
+            {"hogql_val_0": "key"},
+        )
+
+    def test_property_groups_optimized_in_comparisons(self) -> None:
```

Comment:
For later: I've noticed in the query log we do `has(['a', 'b'], properties.key)` fairly often, but those don't seem to be optimized using the bloom filter index. We might want to rewrite those as well here at some point.